### PR TITLE
chore: Switch to using Node built-in test runner

### LINKED ISSUE #2379: Update CI to use `node --test` as the test runner
As outlined in #2373, we need to update our CI configuration to run our tests through `node --test`. Some things that will likely need to be covered by this ticket:

1. Ensure coverage reports are generated such that the CodeCov integration continues to work.
2. Ensure that test file glob matching works on Windows.

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR resolves #2379. It must target the `next` branch because Node.js 16 does not include the test runner.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -161,22 +161,22 @@
     "bench": "node ./bin/run-bench.js",
     "docker-env": "./bin/docker-env-vars.sh",
     "docs": "rm -rf ./out && jsdoc -c ./jsdoc-conf.jsonc --private -r .",
-    "integration": "npm run prepare-test && npm run sub-install && time c8 -o ./coverage/integration tap --test-regex='(\\/|^test\\/integration\\/.*\\.tap\\.js)$' --timeout=600 --no-coverage --reporter classic",
-    "integration:esm": "time c8 -o ./coverage/integration-esm tap --node-arg='--loader=./esm-loader.mjs' --test-regex='(test\\/integration\\/.*\\.tap\\.mjs)$' --timeout=600 --no-coverage --reporter classic",
+    "integration": "npm run prepare-test && npm run sub-install && time c8 -o ./coverage/integration node --test test/integration/**/*.tap.js",
```
Comment (jsumners-nr) [inline package.json:164]: I don't like that we have to rely on shell expansion of the globs in all of these scripts. But we don't run CI on Windows, and in-process expansion of the globs isn't supported until Node 22. So it is what it is.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
I have notice the default reporter `spec` could pose problems in CI.  It does not appear there is an equivalent to tap's `classic` reporter.  If we change to `dot` reporter it will reduce I/O and seems to speed up the test runs as well. This may have to be a follow up because `--test-reporter` doesn't exist until 18

I did suggest some changes as I noticed the globs didn't account for tests in the direct child of a given test type. However, it also appears my proposed changes to include all tests by chaining globs segfaults in node 16, and does not work in 18 and 20.

I did compare the coverage is still the same, as expected because we're still relying on v8 coverage and c8 to report it.

File: package.json [hunk 2]
```diff
@@ -161,22 +161,22 @@
...
     "services:stop": "docker compose down",
-    "smoke": "npm run ssl && time tap test/smoke/**/**/*.tap.js --timeout=180 --no-coverage",
+    "smoke": "npm run ssl && time node --test test/smoke/**/*.tap.js",
     "ssl": "./bin/ssl.sh",
     "sub-install": "node test/bin/install_sub_deps",
     "test": "npm run integration && npm run unit",
     "third-party-updates": "oss third-party manifest --includeOptDeps && oss third-party notices --includeOptDeps && git add THIRD_PARTY_NOTICES.md third_party_manifest.json",
-    "unit": "rm -f newrelic_agent.log && time c8 -o ./coverage/unit tap --test-regex='(\\/|^test\\/unit\\/.*\\.test\\.js)$' --timeout=180 --no-coverage --reporter classic",
-    "unit:scripts": "time c8 -o ./coverage/scripts-unit tap --test-regex='(\\/|^bin\\/test\\/.*\\.test\\.js)$' --timeout=180 --no-coverage --reporter classic",
+    "unit": "rm -f newrelic_agent.log && time c8 -o ./coverage/unit node --test test/unit/**/*.test.js",
```
[the remark this answers is not in the document]
Comment (jsumners-nr) [reply package.json:178]: The subtle difference between the package `glob` doing expansion and the shell doing expansion 😔

**Comment by jsumners-nr** [comment]
> I have notice the default reporter `spec` could pose problems in CI.

What problems might those be?

> If we change to `dot` reporter it will reduce I/O and seems to speed up the test runs as well.

I don't particularly like the `dot` reporter because it is very cryptic. But putting the speed theory to the test:

```
# Unit test run without specifying a reporter:
ℹ tests 176
ℹ suites 0
ℹ pass 176
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 31175.115292

real    0m34.861s
user    1m25.072s
sys     0m31.439s
```

```
# With the dot reporter:
................................................................................................................................................................................

real    0m29.611s
user    1m25.536s
sys     0m32.046s
```

I've done the test multiple times and either reporter is roughly the same amount of time. 🤷‍♂️ I'm okay with either way, but would prefer the standard reporter if it won't cause issues.

**Comment by jsumners-nr** [comment]
> I think that is really useless. I don't enjoy having to search all of the output of the standard reporter, but at least we can see which test(s) failed.

🤔 we could report to `stdout` with the dot reporter but report to something like `unit.txt` via stderr with the spec reporter.

We could also write a simple reporter that outputs test failures when they occur, and nothing on success. In other words, mimic our versioned tests output. Example:

```sh
❯ npm run unit

> newrelic@11.23.1 unit
> rm -f newrelic_agent.log && time c8 -o ./coverage/unit node --test-reporter ./reporter.js --test test/unit/*.test.js test/unit/**/*.test.js

failed: /Users/jsumners/Projects/team-repos/node-newrelic/test/unit/attributes.test.js
real    0m26.942s
user    1m23.630s
sys     0m31.549s
```

I think we could write a much better reporter that shows offending line numbers and such, but the runner in v18 seems to omit a lot of that in my test case (just reporting `line = 1; column = 1` for the induced failure when it should at least be line 35). But I think that report is enough to give us more to go on than the dot reporter and doesn't spam `stdout` with difficult to sift through text.

File: test/lib/test-reporter.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// outputs the failing test file when any failing test has occurred.
+//
+// See https://nodejs.org/api/test.html#custom-reporters.
+'use strict'
+
+const { Transform } = require('node:stream')
+const testReporter = new Transform({
+  writableObjectMode: true,
+  transform(event, encoding, callback) {
+    if (event.type !== 'test:fail') {
```
Comment (bizob2828) [inline test/lib/test-reporter.js:19]: I don't think this is working.  I enhanced it to output the file name on successes and see nothing written to stdout.  If we want to use a custom reporter, i'd still like to see progress and obv failures.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
the update to globs still isn't running all the tests. i found in node 22 if you wrap each glob in single quotes it works. but that fails in 18 and 20

File: test/lib/test-reporter.js (hunk 3 above)
Comment (jsumners-nr) [reply test/lib/test-reporter.js:19]: What change did you make?

Comment (jsumners-nr) [reply test/lib/test-reporter.js:19]: Also, what issues were you originally referring to? I assumed it was the constant output being rendered to the CI log, but now you're stating you want to see that?

File: test/unit/attributes.test.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ tap.test('#addAttribute', (t) => {
     inst.addAttribute(DESTINATIONS.TRANS_SCOPE, 'test', 'success')
     const attributes = inst.get(DESTINATIONS.TRANS_SCOPE)
 
-    t.equal(attributes.test, 'success')
+    t.equal(attributes.test, 'successssss')
```
Comment (bizob2828) [inline test/unit/attributes.test.js]: revert

File: package.json [hunk 5]
```diff
@@ -161,22 +161,22 @@
...
     "services:stop": "docker compose down",
-    "smoke": "npm run ssl && time tap test/smoke/**/**/*.tap.js --timeout=180 --no-coverage",
+    "smoke": "npm run ssl && time borp --reporter ./test/lib/test-reporter.mjs 'test/smoke/**/*.tap.js'",
     "ssl": "./bin/ssl.sh",
     "sub-install": "node test/bin/install_sub_deps",
     "test": "npm run integration && npm run unit",
     "third-party-updates": "oss third-party manifest --includeOptDeps && oss third-party notices --includeOptDeps && git add THIRD_PARTY_NOTICES.md third_party_manifest.json",
-    "unit": "rm -f newrelic_agent.log && time c8 -o ./coverage/unit tap --test-regex='(\\/|^test\\/unit\\/.*\\.test\\.js)$' --timeout=180 --no-coverage --reporter classic",
-    "unit:scripts": "time c8 -o ./coverage/scripts-unit tap --test-regex='(\\/|^bin\\/test\\/.*\\.test\\.js)$' --timeout=180 --no-coverage --reporter classic",
+    "unit": "rm -f newrelic_agent.log && time c8 -o ./coverage/unit borp --reporter ./test/lib/test-reporter.mjs 'test/unit/**/*.test.js'",
```
Comment (bizob2828) [inline package.json:178]: i think a timeout has to be set. the facts.test.js fails

File: package.json [hunk 6]
```diff
@@ -161,22 +161,22 @@
...
+    "integration:esm": "NODE_OPTIONS='--loader=./esm-loader.mjs' time c8 -o ./coverage/integration-esm borp --reporter ./test/lib/test-reporter.mjs 'test/integration/**/*.tap.mjs'",
     "prepare-test": "npm run ssl && npm run docker-env",
     "lint": "eslint ./*.{js,mjs} lib test bin examples",
     "lint:fix": "eslint --fix, ./*.{js,mjs} lib test bin examples",
     "public-docs": "jsdoc -c ./jsdoc-conf.jsonc && cp examples/shim/*.png out/",
     "publish-docs": "./bin/publish-docs.sh",
     "services": "DOCKER_PLATFORM=linux/$(uname -m) docker compose up -d --wait",
     "services:stop": "docker compose down",
-    "smoke": "npm run ssl && time tap test/smoke/**/**/*.tap.js --timeout=180 --no-coverage",
+    "smoke": "npm run ssl && time borp --reporter ./test/lib/test-reporter.mjs 'test/smoke/**/*.tap.js'",
```
Comment (bizob2828) [inline package.json:173]: i think we need to restore the timeout. i have 1 test that fails

File: test/lib/test-reporter.mjs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+// information to be usable in CI, and the `dot` reporter hides which tests
+// failed. This custom reporter outputs nothing for successful tests, and
+// outputs the failing test file when any failing test has occurred.
+//
+// See https://nodejs.org/api/test.html#custom-reporters.
+
+const OUTPUT_MODE = process.env.OUTPUT_MODE?.toLowerCase() ?? 'simple'
+const isSilent = OUTPUT_MODE === 'quiet' || OUTPUT_MODE === 'silent'
+
+async function* reporter(source) {
```
Comment (bizob2828) [inline test/lib/test-reporter.mjs:17]: very nice. we can always add to this.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
overall looks great. i do have some failing tests that i suspect are because of time outs though

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**